### PR TITLE
Add SSL root certificate support

### PR DESCRIPTION
**Pull request by comphead**

### CODE REVIEW

**Comment by comphead** [comment]
@nooberfsh please check the PR

**Comment by nooberfsh** [comment]
> Im wondering do we really need a thin wrapper over reqwest::Certificate?

Yes, we should avoid to public an API which depend on a third party library which may cause version conflict problem.

 I still don't like the `ssl` API, it add another layer to construct the `Client`,  we can just make a module which contains all `ssl` related stuff, e.g.  root_certificate, client certificate, ssl trustore, ssl trust store pwd. [reqwest tls](https://docs.rs/reqwest/latest/reqwest/tls/index.html) is a good example.

File: src/ssl.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    }
+}
+
+impl Ssl {
+    pub fn cert_from_pem<P: AsRef<Path>>(root_certificate_path: P) -> Result<reqwest::Certificate> {
+        let mut buf = Vec::new();
+        std::fs::File::open(root_certificate_path.as_ref())
+            .unwrap()
+            .read_to_end(&mut buf)
+            .unwrap();
```
Comment (nooberfsh) [inline src/ssl.rs:25]: we should really avoid `unwrap`.

File: src/ssl.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        std::fs::File::open(root_certificate_path)
+            .expect(&format!(
+                "Error opening certificate file {}",
+                root_certificate_path
+            ))
+            .read_to_end(&mut buf)
+            .expect(&format!(
+                "Error reading certificate file {}",
+                root_certificate_path
+            ));
```
Comment (nooberfsh) [inline src/ssl.rs:43]: we should really avoid `expect`.

File: src/ssl.rs (hunk 1 above)
Comment (comphead) [reply src/ssl.rs]: Done

File: src/ssl.rs (hunk 2 above)
Comment (comphead) [reply src/ssl.rs]: Done

**Comment by comphead** [comment]
> > Im wondering do we really need a thin wrapper over reqwest::Certificate?
> 
> Yes, we should avoid to public an API which depend on a third party library which may cause version conflict problem.
> 
> I still don't like the `ssl` API, it add another layer to construct the `Client`, we can just make a module which contains all `ssl` related stuff, e.g. root_certificate, client certificate, ssl trustore, ssl trust store pwd. [reqwest tls](https://docs.rs/reqwest/latest/reqwest/tls/index.html) is a good example.

Done with the wrapper.

Reg to ssl module, do you have already an example of such module?

File: src/ssl.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+impl Default for Ssl {
+    fn default() -> Self {
+        Ssl {
+            root_cert: None,
+        }
+    }
+}
+
+impl Ssl {
+    pub fn read_pem<P: AsRef<Path> + std::fmt::Display>(root_certificate_path: &P) -> Result<Certificate> {
```
Comment (nooberfsh) [inline src/ssl.rs:22]: `Display` can be removed as `Path` has [display](https://doc.rust-lang.org/std/path/struct.Path.html#method.display).

File: src/ssl.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        let buf = Self::read_file(&root_certificate_path)?;
+        match reqwest::Certificate::from_pem(&buf) {
+            Ok(cert) => Ok(Certificate { 0: cert }),
+            Err(e) => Err(crate::error::Error::InternalError(format!("Cannot load PEM certificate {:?}", e)))
+        }
+    }
+
+    pub fn read_der<P: AsRef<Path> + std::fmt::Display>(root_certificate_path: &P) -> Result<Certificate> {
```
Comment (nooberfsh) [inline src/ssl.rs:30]: `Display` can be removed as `Path` has [display](https://doc.rust-lang.org/std/path/struct.Path.html#method.display).

File: src/ssl.rs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        let buf = Self::read_file(&root_certificate_path)?;
+        match reqwest::Certificate::from_der(&buf) {
+            Ok(cert) => Ok(Certificate { 0: cert }),
+            Err(e) => Err(crate::error::Error::InternalError(format!("Cannot load DER certificate {:?}", e)))
+        }
+    }
+
+    pub fn read_file<P: AsRef<Path> + std::fmt::Display>(file_path: &P) -> Result<Vec<u8>> {
```
Comment (nooberfsh) [inline src/ssl.rs:38]: We don't need make it public since it's an internal API

File: src/ssl.rs (hunk 4 above)
Comment (comphead) [reply src/ssl.rs]: Done.

File: src/ssl.rs (hunk 3 above)
Comment (comphead) [reply src/ssl.rs]: Done.

**Pull request merged**